File: services/document-ai/service_client.py
"""Inter-service HTTP clients for the Document AI service.

Provides typed Python clients for:
- Auth service (token validation for protected endpoints)
- Election service (form template lookup)
- Compliance service (audit logging for document processing)
- Middleware service (event publishing for processed documents)
"""

import os
import logging
from typing import Any

import httpx

logger = logging.getLogger(__name__)

DEFAULT_TIMEOUT = 10.0


class ServiceClient:
    """Base HTTP client for inter-service communication."""

    def __init__(self, service_name: str, base_url: str):
        self.service_name = service_name
        self.base_url = base_url.rstrip("/")
        self._client = httpx.AsyncClient(
            base_url=self.base_url,
            timeout=DEFAULT_TIMEOUT,
            headers={"X-Service-Name": service_name},
        )

    async def get(self, path: str) -> Any:
        resp = await self._client.get(path)
        resp.raise_for_status()
        return resp.json()

    async def post(self, path: str, body: Any) -> Any:
        resp = await self._client.post(path, json=body)
        resp.raise_for_status()
        return resp.json()

    async def health(self) -> bool:
        try:
            await self.get("/health")
            return True
        except Exception as e:
            logger.warning("%s health check failed: %s", self.service_name, e)
            return False

    async def close(self):
        await self._client.aclose()


class AuthServiceClient(ServiceClient):
    def __init__(self, base_url: str | None = None):
        url = base_url or os.getenv("AUTH_URL", "http://localhost:8090")
        super().__init__("auth-svc", url)

    async def validate_token(self, token: str) -> dict:
        resp = await self._client.get(
            "/auth/me", headers={"Authorization": f"Bearer {token}"}
        )
        resp.raise_for_status()
        return resp.json()


class ElectionServiceClient(ServiceClient):
    def __init__(self, base_url: str | None = None):
        url = base_url or os.getenv("ELECTION_URL", "http://localhost:8091")
        super().__init__("election-svc", url)

    async def get_form_template(self, form_type: str) -> dict:
        return await self.get(f"/elections/forms/{form_type}")


class ComplianceClient(ServiceClient):
    def __init__(self, base_url: str | None = None):
        url = base_url or os.getenv("COMPLIANCE_URL", "http://localhost:8094")
        super().__init__("compliance-svc", url)

    async def log_document_processing(self, doc_id: str, action: str, result: str) -> dict:
        return await self.post(
            "/compliance/audit",
            {
                "action": action,
                "entity_type": "document",
                "entity_id": doc_id,
                "result": result,
            },
        )


class MiddlewareClient(ServiceClient):
    def __init__(self, base_url: str | None = None):
        url = base_url or os.getenv("MIDDLEWARE_URL", "http://localhost:8085")
        super().__init__("middleware-svc", url)

    async def publish_event(self, topic: str, key: str, value: Any) -> dict:
        return await self.post(
            "/kafka/publish", {"topic": topic, "key": key, "value": value}
        )


class ServiceRegistry:
    """Holds typed clients for all services Document AI talks to."""

    def __init__(self):
        self.auth = AuthServiceClient()
        self.election = ElectionServiceClient()
        self.compliance = ComplianceClient()
        self.middleware = MiddlewareClient()

    async def close(self):
        for client in [self.auth, self.election, self.compliance, self.middleware]:
            await client.close()

    async def health_check(self) -> dict[str, bool]:
        import asyncio
        checks = {
            "auth": self.auth.health(),
            "election": self.election.health(),
            "compliance": self.compliance.health(),
            "middleware": self.middleware.health(),
        }
        results = await asyncio.gather(*checks.values(), return_exceptions=True)
        return {name: (r is True) for name, r in zip(checks.keys(), results)}

File: services/inference-engine/src/service_client.rs
//! Inter-service HTTP clients for the INEC microservice mesh.
//! Allows the inference engine to call other services (auth validation,
//! analytics data, event publishing) via typed Rust clients.

use anyhow::{Context, Result};
use reqwest::Client;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use std::time::Duration;
use tracing::warn;

#[derive(Clone)]
pub struct ServiceClient {
    client: Client,
    base_url: String,
    service_name: String,
}

impl ServiceClient {
    pub fn new(service_name: &str, base_url: &str) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(10))
            .build()
            .expect("failed to build HTTP client");
        Self {
            client,
            base_url: base_url.trim_end_matches('/').to_string(),
            service_name: service_name.to_string(),
        }
    }

    pub async fn get<T: DeserializeOwned>(&self, path: &str) -> Result<T> {
        let url = format!("{}{}", self.base_url, path);
        let resp = self
            .client
            .get(&url)
            .header("X-Service-Name", &self.service_name)
            .send()
            .await
            .with_context(|| format!("{}: GET {} failed", self.service_name, path))?;
        let status = resp.status();
        if !status.is_success() {
            anyhow::bail!("{}: GET {} returned {}", self.service_name, path, status);
        }
        resp.json().await.with_context(|| format!("{}: decode failed", self.service_name))
    }

    pub async fn post<B: Serialize, T: DeserializeOwned>(&self, path: &str, body: &B) -> Result<T> {
        let url = format!("{}{}", self.base_url, path);
        let resp = self
            .client
            .post(&url)
            .header("X-Service-Name", &self.service_name)
            .json(body)
            .send()
            .await
            .with_context(|| format!("{}: POST {} failed", self.service_name, path))?;
        let status = resp.status();
        if !status.is_success() {
            anyhow::bail!("{}: POST {} returned {}", self.service_name, path, status);
        }
        resp.json().await.with_context(|| format!("{}: decode failed", self.service_name))
    }

    pub async fn health(&self) -> bool {
        match self.get::<serde_json::Value>("/health").await {
            Ok(_) => true,
            Err(e) => {
                warn!("{} health check failed: {}", self.service_name, e);
                false
            }
        }
    }
}

// --- Typed Service Clients ---

#[derive(Clone)]
pub struct AuthServiceClient(ServiceClient);

impl AuthServiceClient {
    pub fn new(base_url: &str) -> Self {
        Self(ServiceClient::new("auth-svc", base_url))
    }

    pub async fn validate_token(&self, token: &str) -> Result<TokenClaims> {
        let url = format!("{}/auth/me", self.0.base_url);
        let resp = self
            .0
            .client
            .get(&url)
            .header("Authorization", format!("Bearer {}", token))
            .send()
            .await?;
        if !resp.status().is_success() {
            anyhow::bail!("token invalid: {}", resp.status());
        }
        Ok(resp.json().await?)
    }

    pub async fn health(&self) -> bool {
        self.0.health().await
    }
}

#[derive(Clone)]
pub struct AnalyticsClient(ServiceClient);

impl AnalyticsClient {
    pub fn new(base_url: &str) -> Self {
        Self(ServiceClient::new("lakehouse-analytics", base_url))
    }

    pub async fn get_election_data(&self, election_id: i64) -> Result<serde_json::Value> {
        self.0
            .get(&format!("/analytics/election/{}", election_id))
            .await
    }

    pub async fn health(&self) -> bool {
        self.0.health().await
    }
}

#[derive(Clone)]
pub struct EventBusClient(ServiceClient);

impl EventBusClient {
    pub fn new(base_url: &str) -> Self {
        Self(ServiceClient::new("middleware-svc", base_url))
    }

    pub async fn publish(&self, topic: &str, key: &str, value: &serde_json::Value) -> Result<()> {
        #[derive(Serialize)]
        struct PublishReq<'a> {
            topic: &'a str,
            key: &'a str,
            value: &'a serde_json::Value,
        }
        let _: serde_json::Value = self
            .0
            .post(
                "/kafka/publish",
                &PublishReq { topic, key, value },
            )
            .await?;
        Ok(())
    }

    pub async fn health(&self) -> bool {
        self.0.health().await
    }
}

#[derive(Clone)]
pub struct GeoServiceClient(ServiceClient);

impl GeoServiceClient {
    pub fn new(base_url: &str) -> Self {
        Self(ServiceClient::new("geo-svc", base_url))
    }

    pub async fn validate_geofence(&self, lat: f64, lng: f64, pu_id: &str) -> Result<GeofenceResult> {
        self.0
            .post(
                "/geo/validate-geofence",
                &serde_json::json!({
                    "latitude": lat,
                    "longitude": lng,
                    "polling_unit_id": pu_id,
                }),
            )
            .await
    }

    pub async fn health(&self) -> bool {
        self.0.health().await
    }
}

// --- Response Types ---

#[derive(Debug, Clone, Deserialize)]
pub struct TokenClaims {
    pub user_id: String,
    pub role: String,
    #[serde(default)]
    pub permissions: Vec<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct GeofenceResult {
    pub valid: bool,
    pub distance_meters: f64,
    pub polling_unit_id: String,
}

// --- Service Registry ---

#[derive(Clone)]
pub struct ServiceRegistry {
    pub auth: AuthServiceClient,
    pub analytics: AnalyticsClient,
    pub event_bus: EventBusClient,
    pub geo: GeoServiceClient,
}

impl ServiceRegistry {
    pub fn from_env() -> Self {
        let auth_url = std::env::var("AUTH_URL").unwrap_or_else(|_| "http://localhost:8090".into());
        let analytics_url =
            std::env::var("LAKEHOUSE_URL").unwrap_or_else(|_| "http://localhost:8098".into());
        let middleware_url =
            std::env::var("MIDDLEWARE_URL").unwrap_or_else(|_| "http://localhost:8085".into());
        let geo_url = std::env::var("GEO_URL").unwrap_or_else(|_| "http://localhost:8093".into());

        Self {
            auth: AuthServiceClient::new(&auth_url),
            analytics: AnalyticsClient::new(&analytics_url),
            event_bus: EventBusClient::new(&middleware_url),
            geo: GeoServiceClient::new(&geo_url),
        }
    }
}
